### PR TITLE
Add comprehensive tests for LLMChatWithParsingRetryBlock

### LINKED ISSUE #346: Add retry mechanism for LLM parsing failures in composite block
## Summary

Add a composite block that combines LLMChatBlock and TextParserBlock with automatic retry logic to handle parsing failures. When the LLM output cannot be parsed by TextParserBlock, the block should automatically retry the LLM call until a parseable result is obtained or a maximum retry limit is reached.

## Problem Description

Currently, when an LLM doesn't provide output in the expected format (e.g., requesting JSON but receiving unparseable text), the workflow fails at the TextParserBlock stage. Users need a robust way to retry LLM calls until the output meets parsing requirements, similar to how existing retry mechanisms handle API errors.

## Proposed Solution

Create a composite block (e.g., `LLMChatWithParsingRetryBlock`) that:

1. **Internally combines**:
   - `LLMChatBlock` for LLM generation 
   - `TextParserBlock` for output parsing

2. **Retry Logic**:
   - Calls `LLMChatBlock.generate()` to get LLM response
   - Attempts to parse using `TextParserBlock.generate()`
   - If parsing fails, retries the LLM call (not just parsing)
   - Continues until parsing succeeds OR max retries reached
   - Returns successfully parsed results or raises appropriate error

3. **Configuration Parameters**:
   - `max_parsing_retries`: Maximum number of retry attempts (default: 3)
   - All existing `LLMChatBlock` parameters for LLM configuration
   - All existing `TextParserBlock` parameters for parsing configuration

## Technical Context

- **Current Architecture**: The framework uses modular blocks (`src/sdg_hub/core/blocks/llm/llm_chat_block.py`, `src/sdg_hub/core/blocks/llm/text_parser_block.py`) that can be composed into flows
- **Existing Composite Patterns**: Modern composite blocks like `EvaluateFaithfulnessBlock` (`src/sdg_hub/core/blocks/evaluation/evaluate_faithfulness_block.py`) and `EvaluateRelevancyBlock` (`src/sdg_hub/core/blocks/evaluation/evaluate_relevancy_block.py`) demonstrate the proper patterns for creating internal blocks, chaining them in `generate()`, and handling validation across the chain
- **Retry Scope**: This is specifically about parsing failures requiring new LLM calls, not API-level retries (which are already handled by tenacity/backoff mechanisms)

## Expected Interface

```yaml
blocks:
  - block_type: "LLMChatWithParsingRetryBlock"
    block_config:
      block_name: "reliable_json_generator"
      input_cols: "messages"
      output_cols: "parsed_json"
      max_parsing_retries: 5
      
      # LLM Configuration
      model: "openai/gpt-4"
      temperature: 0.7
      response_format: {"type": "json_object"}  # Supported in LLMChatBlock
      
      # Parser Configuration  
      parsing_pattern: "\\{.*\\}"  # Extract JSON objects
      parser_cleanup_tags: ["```json", "```"]
```

## Benefits

- **Improved Reliability**: Reduces workflow failures due to format inconsistencies
- **User Experience**: Automatic handling of common LLM output format issues
- **Composability**: Maintains the framework's modular design principles
- **Backward Compatibility**: Doesn't affect existing block implementations

## Implementation Notes

- Follow existing composite block patterns from `EvaluateFaithfulnessBlock` and `EvaluateRelevancyBlock`
- Use `_create_internal_blocks()` method to instantiate internal blocks
- Chain blocks in `generate()` method with proper error handling for retry logic
- Ensure proper error handling and logging throughout retry cycles
- Consider adding metrics/logging for retry attempts and success rates
- Maintain thread safety for async operations

### PR DESCRIPTION
**Pull request by eshwarprasadS**
Fixes #346 

## Summary

- Added comprehensive test suite for `LLMChatWithParsingRetryBlock` composite block
- 31 tests covering all aspects of the retry logic and internal block coordination
- Tests validate both tag-based and regex-based parsing with retry functionality
- Full coverage of error scenarios, edge cases, and integration behavior

### Retry Logic & Error Handling
- **Parsing retry mechanism**: Automatically retries LLM generation when parsing fails
- **Result accumulation**: Collects successful parses across retry attempts until target count is reached
- **MaxRetriesExceededError**: Proper error handling when target count isn't achieved within max retries
- **Partial success scenarios**: Handles cases where some but not all responses parse successfully

### Internal Block Coordination  
- **LLMChatBlock integration**: Seamless parameter passing to internal LLM block
- **TextParserBlock integration**: Proper configuration of parsing rules and cleanup tags
- **Parameter validation**: Ensures all internal blocks are properly configured
- **Error propagation**: Correctly handles and propagates errors from internal blocks

### Configuration & Validation
- **Both parsing methods**: Tag-based (`<tag>content</tag>`) and regex-based parsing
- **Multiple output columns**: Support for parsing multiple fields from single response
- **LLM parameter passthrough**: All generation parameters (temperature, max_tokens, n, etc.)
- **Input validation**: Proper validation of parsing configuration and model requirements

### Advanced Scenarios
- **Multiple responses (n > 1)**: Generates and parses multiple completions per input
- **Result trimming**: Automatically trims excess results when target count is exceeded
- **Async mode support**: Validates async configuration is passed to internal blocks
- **Runtime parameter overrides**: Supports overriding parameters at generation time

## Example Usage

```python
# Example test demonstrating retry logic
def test_successful_generation_after_retry(self, sample_dataset):
    """Test successful generation after initial parsing failures."""
    with patch("sdg_hub.core.blocks.llm.client_manager.completion") as mock_completion:
        # First call returns unparseable, second returns parseable
        mock_response_bad = MagicMock()
        mock_response_bad.choices[0].message.content = "No tags here"
        
        mock_response_good = MagicMock()
        mock_response_good.choices[0].message.content = "<answer>Good response</answer>"
        
        # Alternate between bad and good responses
        mock_completion.side_effect = [
            mock_response_bad, mock_response_good,  # For first sample
            mock_response_bad, mock_response_good,  # For second sample
        ]

        block = LLMChatWithParsingRetryBlock(
            block_name="test_retry_success",
            input_cols="messages",
            output_cols="answer",
            model="openai/gpt-4",
            start_tags=["<answer>"],
            end_tags=["</answer>"],
            parsing_max_retries=3,
        )

        result = block.generate(sample_dataset)

        # Should succeed after retry
        assert len(result) == 2
        assert all(row["answer"] == "Good response" for row in result)
        
        # Should have called LLM twice per sample (1 retry each)
        assert mock_completion.call_count == 4
```

## Usage Example

```python
# Tag-based parsing with retry
block = LLMChatWithParsingRetryBlock(
    block_name="qa_with_retry",
    input_cols="messages",
    output_cols=["explanation", "answer"],
    model="openai/gpt-4",
    start_tags=["<explanation>", "<answer>"], 
    end_tags=["</explanation>", "</answer>"],
    parsing_max_retries=3,
    n=2  # Generate 2 responses per input
)

# Regex-based parsing with retry
block = LLMChatWithParsingRetryBlock(
    block_name="json_with_retry",
    input_cols="messages", 
    output_cols="result",
    model="anthropic/claude-3-sonnet-20240229",
    parsing_pattern=r'"result":\s*"([^"]*)"',
    parsing_max_retries=5
)
```

## Test Coverage

- ✅ **All 157 LLM block tests pass** (31 new + 126 existing)
- ✅ **No regressions** in existing functionality  
- ✅ **Comprehensive error scenarios** covered
- ✅ **Edge cases** and **integration behavior** validated
- ✅ **Both parsing methods** (tags and regex) tested
- ✅ **Proper mocking** following existing test patterns

## Test plan

- [x] Run the full test suite: `pytest tests/blocks/llm/ -v`
- [x] Verify all 157 tests pass without regressions
- [x] Check test coverage for the new composite block
- [x] Validate test patterns match existing codebase standards

🤖 Generated with [Claude Code](https://claude.ai/code)

## Summary by CodeRabbit

- New Features
  - Introduced an LLM chat block with automatic parsing retries for more reliable structured outputs.
  - Supports configurable retry limits, tag- or regex-based parsing, and trims results to requested counts.
  - Exposes rich generation settings (e.g., temperature, max tokens, sampling, streaming) and is now available in the public API.

- Tests
  - Added comprehensive tests covering initialization, configuration validation, retry behavior, error handling, integration, and internal status reporting.

### CODE REVIEW

File: src/sdg_hub/core/blocks/llm/llm_chat_with_parsing_retry_block.py [hunk 1]
```diff
@@ -0,0 +1,569 @@
...
+    "Composite block combining LLM chat and text parsing with automatic retry on parsing failures",
+)
+class LLMChatWithParsingRetryBlock(BaseBlock):
+    """Composite block for LLM generation with parsing retry logic.
+
+    This block combines LLMChatBlock and TextParserBlock into a single cohesive block
+    that automatically retries LLM generation when parsing fails, accumulating successful
+    results until the target count is reached or max retries exceeded.
+
+    Parameters
```
Comment (shivchander) [inline src/sdg_hub/core/blocks/llm/llm_chat_with_parsing_retry_block.py:52]: I have one architectural suggestion that could significantly simplify the code.

Suggestion: Eliminate Parameter Duplication

Currently, the block redefines all 20+ parameters from LLMChatBlock and TextParserBlock (lines 149-220), which creates maintenance overhead and code duplication. Consider using dynamic parameter forwarding instead - this is the standard pattern used by frameworks like Django, scikit-learn, and TensorFlow.

Proposed approach:

```python
class LLMChatWithParsingRetryBlock(BaseBlock):
    # Only define composite-specific parameters
    parsing_max_retries: int = Field(3, description="Max retry attempts for parsing failures")

    def __init__(self, **kwargs):
        # Extract composite-specific params
        self.parsing_max_retries = kwargs.pop('parsing_max_retries', 3)

        # Forward parameters to appropriate internal blocks
        llm_params = {k: v for k, v in kwargs.items()
                     if k in LLMChatBlock.model_fields}
        parser_params = {k: v for k, v in kwargs.items()
                        if k in TextParserBlock.model_fields}

        super().__init__(**{k: v for k, v in kwargs.items()
                           if k in BaseBlock.model_fields})

        self._create_internal_blocks(llm_params, parser_params)
```

Comment (eshwarprasadS) [reply src/sdg_hub/core/blocks/llm/llm_chat_with_parsing_retry_block.py:52]: addressed this in the new commit. The pattern may not fit 1:1 for pydantic models (opposed to base python classes) but I have adopted something very similar to the suggested pattern, removing the core problem of duplicate parameters and associated management and memory overheads.
